package cloudasr.messages;

message HeartbeatMessage {
	required string address = 1;
	required string model = 2;
	required Status status = 3;

	enum Status {
		STARTED = 0;
		WAITING = 1;
		WORKING = 2;
		FINISHED = 3;
	};
}

message WorkerStatusMessage {
	required string address = 1;
	required string model = 2;
	required Status status = 3;
	required int32 time = 4;

	enum Status {
		STARTED = 0;
		WAITING = 1;
		WORKING = 2;
	};
}

message WorkerRequestMessage {
	required string model = 1;
}

message MasterResponseMessage {
	required Status status = 1;
	optional string address = 2;

	enum Status {
		SUCCESS = 0;
		ERROR = 1;
	};
}

message RecognitionRequestMessage {
	required UniqueID id = 1;
	required Type type = 2;
	required bytes body = 3;
	optional bool has_next = 4;
	optional int32 frame_rate = 5;

	enum Type {
		BATCH = 0;
		ONLINE = 1;
	}
}

message Alternative {
	required string transcript = 1;
	required float confidence = 2;
}

message ResultsListMessage {
	repeated ResultsMessage results = 1;
}

message ResultsMessage {
	required Status status = 1;
	optional UniqueID id = 2;
	optional bool final = 3;
	repeated Alternative alternatives = 4;

	enum Status {
		SUCCESS = 0;
		ERROR = 1;
	};
}

message UniqueID {
	required uint64 lower = 1;
	required uint64 upper = 2;
}

message SaverMessage {
	required UniqueID id = 1;
	required int32 part = 2;
	required UniqueID chunk_id = 3;
	required string model = 4;
	required bytes body = 5;
	required int32 frame_rate = 6;
	repeated Alternative alternatives = 7;
}
